### PR TITLE
fix(materialized columns): Create materialized columns without dollar sign in column name

### PR DESCRIPTION
**Pull request by macobo**
## Changes

Follow-up to https://github.com/PostHog/posthog/pull/8846

Our materialization logic currently creates columns with a $ sign. According to the linked PR, clickhouse added heredoc support, breaking this feature.

This PR introduces an alternative solution by removing $ from the column names. I'd love to roll back the linked PR and use this instead as I think this has less of a chance of becoming a frequent footgun during development.

## How did you test this code?

See tests. Also ran the migration on an existing install as well as a fresh one.

Planning on manually testing the migration against prod before merging.

### CODE REVIEW

File: ee/clickhouse/models/event.py
Comment (macobo) [inline ee/clickhouse/models/event.py:231]: Note: This was a spot missed in the previous PR

Comment (hazzadous) [reply ee/clickhouse/models/event.py:231]: Yes I did just enough to fix the tests an no more, leaving that to a separate followup.

**Comment by hazzadous** [comment]
> I'd love to roll back the linked PR and use this instead as I think this has less of a chance of becoming a frequent footgun during development

I'd suggest we do both, if we can ensure that all identifiers are quoted then that should remove some footguns, coupled with only using boring characters in identifier names.

File: ee/clickhouse/queries/paths/path_event_query.py
Comment (macobo) [inline ee/clickhouse/queries/paths/path_event_query.py:51]: Also needed to clean this up due to a potential SQL injection risk/usage of $ in column names.

File: ee/clickhouse/migrations/0025_remove_dollar_from_materialized_columns.py
Comment (hazzadous) [inline ee/clickhouse/migrations/0025_remove_dollar_from_materialized_columns.py:29]: Is there any follow up needed in this case, for the operator?

File: ee/clickhouse/migrations/0025_remove_dollar_from_materialized_columns.py
Comment (hazzadous) [inline ee/clickhouse/migrations/0025_remove_dollar_from_materialized_columns.py:30]: This seems more like a warning rather than info. Also, can we capture the exception with `exc_info=True`.

File: ee/clickhouse/migrations/0025_remove_dollar_from_materialized_columns.py
Comment (hazzadous) [inline ee/clickhouse/migrations/0025_remove_dollar_from_materialized_columns.py:33]: What happens while this is running, or if we fail part way through for some reason? If old code is still trying to access $group will that be an issue?

**Review by hazzadous — APPROVED** [review]
Looks good, a bit hazy on the exact details.

Looks like there was still a reference to $ in `ee/clickhouse/queries/test/__snapshots__/test_stickiness.ambr` when I checked, maybe this is a stale snapshot?

File: ee/clickhouse/migrations/0025_remove_dollar_from_materialized_columns.py
Comment (macobo) [reply ee/clickhouse/migrations/0025_remove_dollar_from_materialized_columns.py:29]: Nope - it's already renamed, can ignore.

File: ee/clickhouse/migrations/0025_remove_dollar_from_materialized_columns.py
Comment (macobo) [reply ee/clickhouse/migrations/0025_remove_dollar_from_materialized_columns.py:30]: Will clear up the wording. This is safe to ignore as the column would already be renamed on a fresh install.

File: ee/clickhouse/migrations/0025_remove_dollar_from_materialized_columns.py
Comment (macobo) [reply ee/clickhouse/migrations/0025_remove_dollar_from_materialized_columns.py:33]: Good question.

We won't fail part way through, but old code will be running even after this is complete. Due to caching a few queries will fail as this is deployed as a result until the new pods are live. As such, I'd deploy this during a quieter timeframe.

File: ee/clickhouse/migrations/0025_remove_dollar_from_materialized_columns.py
Comment (rcmarron) [inline ee/clickhouse/migrations/0025_remove_dollar_from_materialized_columns.py:45]: Flyby comment: if a self-hosted customer had `$mat_session_id` materialized on their instance, could this fail bc we're trying to rename `$mat_session_id` to `mat_session_id` which just got created above?

Comment (macobo) [reply ee/clickhouse/migrations/0025_remove_dollar_from_materialized_columns.py:45]: Might be because I guess the migration creating the column `mat_session_id` broke the invariant that one property is only materialized once :(

Comment (macobo) [reply ee/clickhouse/migrations/0025_remove_dollar_from_materialized_columns.py:45]: Nevermind, we validate that invariant in code.

**Review by guidoiaquinti — APPROVED** [review]
Merge conflict but overall LGTM

**Comment by hazzadous** [comment]
@macobo to let you get on with other tasks did you want to pass this resolution along to someone else? Is there anything else to do here other than resolve conflicts?

**Comment by macobo** [comment]
> @macobo to let you get on with other tasks did you want to pass this resolution along to someone else? Is there anything else to do here other than resolve conflicts?

Nope, I will get this merged. This was originally blocked by the trouble we were having on ch8, now want to make sure it works well with sharded setups and depends on https://github.com/PostHog/posthog/pull/9013.

**Comment by macobo** [comment]
Deploying first thing tomorrow...

File: ee/clickhouse/migrations/0025_remove_dollar_from_materialized_columns.py
Comment (hazzadous) [inline ee/clickhouse/migrations/0025_remove_dollar_from_materialized_columns.py:33]: We do this if and the IF NOT EXIST SQL to avoid hitting the db all the time, is that right?

File: ee/clickhouse/migrations/0025_remove_dollar_from_materialized_columns.py
Comment (hazzadous) [inline ee/clickhouse/migrations/0025_remove_dollar_from_materialized_columns.py:19]: Is it always varchar? I guess we treat everything as a string then interpret on read if we need to?

Comment (macobo) [reply ee/clickhouse/migrations/0025_remove_dollar_from_materialized_columns.py:19]: Materialized columns are all varchars currently yeah. :)

File: ee/clickhouse/migrations/0025_remove_dollar_from_materialized_columns.py
Comment (macobo) [reply ee/clickhouse/migrations/0025_remove_dollar_from_materialized_columns.py:33]: IF NOT EXISTS is for checking if the _new_ column exists, this method is for checking whether the _old_ column also exists (which it won't on new setups)

Comment (hazzadous) [reply ee/clickhouse/migrations/0025_remove_dollar_from_materialized_columns.py:33]: Ah I see it's for the current name, ignore

File: ee/clickhouse/migrations/0025_remove_dollar_from_materialized_columns.py
Comment (hazzadous) [inline ee/clickhouse/migrations/0025_remove_dollar_from_materialized_columns.py:25]: If we fail for any reason on the second comment, I guess we would not used a materialized column for some queries. Is it worth switching the order here such that we always maintain at least one materialized column comment (I have no idea what the behaviour here is however, but maybe it just uses the first one it comes across?)

I assume also that it will use the cached materialized column names also, but that's fine because the old name still exists. I like this solution

Comment (macobo) [reply ee/clickhouse/migrations/0025_remove_dollar_from_materialized_columns.py:25]: > Is it worth switching the order here such that we always maintain at least one materialized column comment 

I am not sure how code would behave if some comments are duplicated - I think it'll be undeterministic. Losing columns sounds better this way.

You're right in that this migration isn't really rollbackable, but it should be fine as prod is currently the only sharded cluster if we take care there.

File: ee/clickhouse/migrations/0025_remove_dollar_from_materialized_columns.py
Comment (hazzadous) [inline ee/clickhouse/migrations/0025_remove_dollar_from_materialized_columns.py:70]: I assume `materialized_column_name` never returns something with a `$` because of https://github.com/PostHog/posthog/pull/8855/files#diff-559ae24ad49a1674573c0c89db08c7b6d96c0920e2625e0877b2155f0f453aa1L141

And that is tested by https://github.com/PostHog/posthog/pull/8855/files#diff-6653caa40d66302b910d83e95b18f8176e45400c230eed900e5dc7ef222577ceL91 ?

**Review by hazzadous — APPROVED** [review]
Looks good 🚀 

I didn't see a test for this but I may have missed. I would love to know that the code has at least run before trying out live, but I leave that to your discretion.

**Comment by macobo** [comment]
> I would love to know that the code has at least run before trying out live, but I leave that to your discretion.

It has! Locally ran migrations on master, created a materialized column, switched here, ran 0025 and smoke-tested querying the old and new columns + what get_materialized_columns returns.

**Comment by fuziontech** [comment]
Revived!

**Comment by fuziontech** [comment]
Closing in favor of https://github.com/PostHog/posthog/pull/9721
Thanks @rcmarron !

**Pull request closed without merging**
